lib/openssl/x509.rb

```diff
@@ -60,6 +60,109 @@ def to_h # {"oid"=>sn|ln, "value"=>value, "critical"=>true|false}
       def to_a
         [ self.oid, self.value, self.critical? ]
       end
+
+      module Helpers
+        def find_extension(oid)
+          extensions.find { |e| e.oid == oid }
+        end
+
+        def x509_name_from_general_name(gn_asn1)
+          unless gn_asn1.tag_class == :CONTEXT_SPECIFIC
+            raise ASN1::ASN1Error "invalid GeneralName"
+          end
+
+          if gn_asn1.tag == 4
+            Name.new(gn_asn1.value.first.to_der)
+          else
+            # TODO: raise error instead of returning nil?
+            # this is some other kind of general name.
+            nil
+          end
+        end
+      end
+
+      module SubjectKeyIdentifier
+        include Helpers
+
+        # Described in RFC5280 Section 4.2.1.2
+        #
+        # Returns a binary String key identifier or nil.
+        def subject_key_identifier
+          ext = find_extension("subjectKeyIdentifier")
+          return nil if ext.nil?
+
+          ski_asn1 = ASN1.decode(ext.value_der)
+          if ext.critical? || ski_asn1.tag_class != :UNIVERSAL || ski_asn1.tag != ASN1::OCTET_STRING
+            raise ASN1::ASN1Error "invalid extension"
+          end
+
+          ski_asn1.value
+        end
+      end
+
+      module AuthorityKeyIdentifier
+        include Helpers
+
+        # Described in RFC5280 Section 4.2.1.1
+        #
+        # Returns a Hash with keys :key_identifier, :authority_cert_issuer,
+        # and :authority_cert_serial_number.
+        def authority_key_identifier
```

Comment:
Looking at [Go's implementation](https://golang.org/pkg/crypto/x509/#Certificate), they only expose the `key_identifier` part of this extension. That seems like a reasonable choice, given that it's what is mostly useful for matching certificates to their issuers. Going that route would also simplify this method here.